### PR TITLE
Interactivity API: Update interactive regions during client-side navigation

### PR DESCRIPTION
**Pull request by DAreRodz**
With this PR, the Interactivity API exposes two directives to ease updating the Products Query block during client-side navigation.

- `data-wc-navigation-id`: defines an ID for an interactive region (i.e., an element with the `data-wc-interactive` attribute). The ID matches the region between different pages and is used to render the SSRed content.
- `data-wc-navigation-link`: formerly `data-wc-link`, just renamed to include the `navigation` namespace.

All the logic related to the client-side navigation meta tag has been removed, as it is not needed from now on.

In addition, I've added a couple of filters to the Products Query block in order to inject the directives mentioned above. You can see it working in this video:

| Before | After |
| ------ | ----- |
| <video src="https://github.com/woocommerce/woocommerce-blocks/assets/6917969/8ec402c1-a723-4e80-a76c-9e9987192a5f"></video> | <video src="https://github.com/woocommerce/woocommerce-blocks/assets/6917969/ebf388ba-c3d7-4f5a-89f7-43a26a041f66"></video> |

Tracking issue: woocommerce/woocommerce#42486

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

1. Ensure that you are using the Blockified Product Catalog Template. If not, enable it.
2. Open the Site Editor and go to the Product Catalog Template.
3. In order to make pagination more noticeable, in the Product query block, disable the inherited query settings and manually modify the block markup to set the `perPage` prop inside the `query` attribute to `1`.
4. Visit the Product Catalog (`/shop`).
5. Click on any link inside the Pagination block.
6. Via dev tools, ensure that only the HTML inside the Query block is updated.
7. Navigate back and forward.
8. Via dev tools, ensure that only the HTML inside the Query block is updated.
9. Click on any link outside the Pagination block
10. Ensure that the browser navigates to the selected link.
* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Add suggested changelog entry here.

### CODE REVIEW

**Comment by DAreRodz** [comment]
Feel free to modify the PR description or add the convenient labels. Thanks! 🙏 

cc: @gigitux, @Aljullu

File: src/BlockTypes/ProductQuery.php
Comment (DAreRodz) [inline src/BlockTypes/ProductQuery.php:126]: I used this check to ensure that we're adding the `navigation-link` directive only to the links inside the Pagination Block contained in the Products Query block. If you think there's a better way to do this, feel free to update the code. 😊

File: assets/js/interactivity/router.js
Comment (luisherranz) [inline assets/js/interactivity/router.js:171]: We should prepopulate the cache of the initial page as we were doing in the full-page implementation. That way, if the user goes back to the original URL, we don't need to fetch it again.

File: src/BlockTypes/ProductQuery.php
Comment (luisherranz) [inline src/BlockTypes/ProductQuery.php:107]: There's not a real need to hydrate the inner content of the Products block yet, is it? Maybe we can just hydrate the pagination for now…

File: src/BlockTypes/ProductQuery.php
Comment (luisherranz) [inline src/BlockTypes/ProductQuery.php]: Maybe we should namespace this, to make it more future-proof, just in case. Something like:

```suggestion
				$p->set_attribute( 'data-wc-navigation-id', 'woo-products-' . $block['attrs']['queryId'] );
```

File: src/BlockTypes/ProductQuery.php
Comment (luisherranz) [inline src/BlockTypes/ProductQuery.php:130]: Taking into account that this is not an experiment anymore, maybe we could start with a prefetch strategy of only prefetching the previous/next links, and skipping the page number links?

cc: @gigitux, @aljullu, any thoughts on the prefetching strategy for the pagination?

File: src/BlockTypes/ProductQuery.php
Comment (luisherranz) [reply src/BlockTypes/ProductQuery.php:126]: Yeah, it'd be great if someone from Woo could review this part and give a thumbs up 🙂

**Review by luisherranz — CHANGES_REQUESTED** [review]
Great work! There are still a few things to review, but this looks mostly ready 🙂

File: src/BlockTypes/ProductQuery.php
Comment (Aljullu) [reply src/BlockTypes/ProductQuery.php:130]: Good question. I don't have a strong opinion, but prefetching the next and previous ones and skip the others makes sense to me.

File: src/BlockTypes/ProductQuery.php
Comment (Aljullu) [inline src/BlockTypes/ProductQuery.php]: I guess it depends on the use-case, but when using the Products block in the Product Catalog template, I think it should default to scrolling to top.

**Review by Aljullu — COMMENTED** [review]
I did a brief review and overall it looks good to me. Good work and exciting stuff! :clap: I will let the others do the proper review as they have more context on the Interactivity API.

One question I have is whether the plan is to merge this so it works with the Products (Beta) block or whether it would be better to wait until the Product Collection block is released. Has this been discussed? For context: the Product Collection block is a replacement for the Products (Beta) block that won't use the Gutenberg Query Loop block underneath.

Besides that, I left a couple of minor comments below. :point_down:

File: src/BlockTypes/ProductQuery.php
Comment (DAreRodz) [reply src/BlockTypes/ProductQuery.php:107]: I'll try and see if it works as expected (I guess so). 👍

File: src/BlockTypes/ProductQuery.php
Comment (DAreRodz) [reply src/BlockTypes/ProductQuery.php]: Oh, you mean to avoid collisions between different blocks that could use the same ID for their `data-wc-navigation-id` attribute, right? That makes sense. 👍

File: assets/js/interactivity/router.js
Comment (DAreRodz) [reply assets/js/interactivity/router.js:171]: Fixed in https://github.com/woocommerce/woocommerce-blocks/pull/10200/commits/35c0e5f0a96062a2e109903fd379bbbae47d515d ✅

File: src/BlockTypes/ProductQuery.php
Comment (DAreRodz) [reply src/BlockTypes/ProductQuery.php:107]: ~It works; fixed in https://github.com/woocommerce/woocommerce-blocks/pull/10200/commits/7b4c991f5d735c511c9d75e3a29be27a96684715~ Edit: It actually **didn't work** 😅

More info below.

File: src/BlockTypes/ProductQuery.php
Comment (DAreRodz) [reply src/BlockTypes/ProductQuery.php]: Fixed in https://github.com/woocommerce/woocommerce-blocks/pull/10200/commits/ade05add4196d7e506c55b433f9a74a8b66f79bd ✅

File: src/BlockTypes/ProductQuery.php
Comment (DAreRodz) [reply src/BlockTypes/ProductQuery.php:130]: I'll make that change.

File: src/BlockTypes/ProductQuery.php
Comment (DAreRodz) [reply src/BlockTypes/ProductQuery.php]: I'll make this change as well. 🙂

File: src/BlockTypes/ProductQuery.php
Comment (DAreRodz) [reply src/BlockTypes/ProductQuery.php:107]: The issue was that Preact was instantiating two different root fragments, one for the Query block and another for the Query Pagination block, thus leading to inconsistencies and issues reconciling the HTML.

Additionally, root fragments were instantiated for every client-side navigation instead of reusing them.

**Comment by DAreRodz** [comment]
I've addressed most of your suggestions. Ready for review again. 😊

File: src/BlockTypes/ProductQuery.php
Comment (luisherranz) [reply src/BlockTypes/ProductQuery.php:107]: Thanks for investigating that possibility, David.

File: src/BlockTypes/ProductQuery.php
Comment (luisherranz) [inline src/BlockTypes/ProductQuery.php:152]: Cool. Time to add support for `data-wc-key` to avoid adding invalid HTML?

**Review by luisherranz — APPROVED** [review]
This looks great, David. Amazing work 👏 

Let's wait for a review from Woo, tho. Especially [this bit](https://github.com/woocommerce/woocommerce-blocks/pull/10200#pullrequestreview-1529879150) and an answer to [these Albert's questions](https://github.com/woocommerce/woocommerce-blocks/pull/10200#pullrequestreview-1536695122).

File: src/BlockTypes/ProductQuery.php
Comment (gigitux) [reply src/BlockTypes/ProductQuery.php:126]: I would add another check:
```php
core/query' === $block['blockName'] && self::is_woocommerce_variation( $this->parsed_block )....
```

The rest LGTM!

File: src/BlockTypes/ProductQuery.php
Comment (gigitux) [inline src/BlockTypes/ProductQuery.php:95]: I would add another check:
```php
core/query' === $block['blockName'] && self::is_woocommerce_variation( $this->parsed_block )
```

The rest LGTM!

**Review by gigitux — APPROVED** [review]
I just added a comment but it isn't a critical change. So LGTM! Nice work!

I want to echo the @Aljullu comment:

> One question I have is whether the plan is to merge this so it works with the Products (Beta) block or whether it would be better to wait until the Product Collection block is released. Has this been discussed? For context: the Product Collection block is a replacement for the Products (Beta) block that won't use the Gutenberg Query Loop block underneath.

I don't think that it is a critical issue porting these changes on the Product Collection block, btw.

cc @imanish003 @kmanijak (they are working on the Product Collection)

**Comment by kmanijak** [comment]
> > One question I have is whether the plan is to merge this so it works with the Products (Beta) block or whether it would be better to wait until the Product Collection block is released. Has this been discussed? For context: the Product Collection block is a replacement for the Products (Beta) block that won't use the Gutenberg Query Loop block underneath.

> I don't think that it is a critical issue porting these changes on the Product Collection block, btw.
> cc @imanish003 @kmanijak (they are working on the Product Collection)

@gigitux, thanks for the ping!

First of all, I love the UX improvements this PR brings! And having said that, I think _it is_ crucial to port these to Product Collection as soon as possible. When migrating from Products (Beta) to Product Collection, users would see a decreased performance, which could be perceived as a regression. I don't think it has to be covered within this PR, but would be great to cover it before or right after the initial release. We'd like to release Product Collection in one of the upcoming releases (1-2).

BTW, I may have a wrong impression, but checking the changes in this PR, it doesn't seem too hard to re-use that in Product Collection (I mean especially for someone experienced with Interactivity API).

**Comment by gigitux** [comment]
> > > One question I have is whether the plan is to merge this so it works with the Products (Beta) block or whether it would be better to wait until the Product Collection block is released. Has this been discussed? For context: the Product Collection block is a replacement for the Products (Beta) block that won't use the Gutenberg Query Loop block underneath.
> 
> > I don't think that it is a critical issue porting these changes on the Product Collection block, btw.
> > cc @imanish003 @kmanijak (they are working on the Product Collection)
> 
> @gigitux, thanks for the ping!
> 
> First of all, I love the UX improvements this PR brings! And having said that, I think _it is_ crucial to port these to Product Collection as soon as possible. When migrating from Products (Beta) to Product Collection, users would see a decreased performance, which could be perceived as a regression. I don't think it has to be covered within this PR, but would be great to cover it before or right after the initial release. We'd like to release Product Collection in one of the upcoming releases (1-2).
> 

This is a good point!

> BTW, I may have a wrong impression, but checking the changes in this PR, it doesn't seem too hard to re-use that in Product Collection (I mean especially for someone experienced with Interactivity API).

I'm happy to help you in the next weeks if you prefer, but I agree with your impression. It should be pretty easy to apply the changes done for the Products block on the Product Collection block.

**Comment by kmanijak** [comment]
> I'm happy to help you in the next weeks if you prefer, but I agree with your impression. It should be pretty easy to apply the changes done for the Products block on the Product Collection block.

Thanks! Issue created: https://github.com/woocommerce/woocommerce-blocks/issues/10328, I'll ping you when I'm going to pick this up 🙌

**Comment by imanish003** [comment]
FYI I'll be working on adding this to the Product Collection block, and I've assigned the issue to myself on https://github.com/woocommerce/woocommerce-blocks/issues/10328. For that, I will checkout a new branch from this PR branch. 

CC: @kmanijak @gigitux

**Comment by imanish003** [comment]
Also, In [woocommerce-gutenberg-products-block.php](https://github.com/woocommerce/woocommerce-blocks/blob/8f1c4a314088d41d329d0441ecc0808214144b24/woocommerce-gutenberg-products-block.php), we have:

> Requires at least: 6.1
> 

But this PR uses `WP_HTML_Tag_Processor`, which was introduced in `6.2.0` as mentioned [here](https://developer.wordpress.org/reference/classes/wp_html_tag_processor/#changelog). I think we will need to do one of the following (unless I am missing something):

1. Update `Requires at least` to `6.2` ,OR
2. Conditionally support the features introduced in this PR , OR
3. Figure out alternative to `WP_HTML_Tag_Processor` to update the $block_content

CC: @tjcafferkey @Aljullu

**Review by imanish003 — CHANGES_REQUESTED** [review]
I noticed an issue when there are two Products beta blocks on the same post. When I click on "Next Page," the value of `data-wc-navigation-id` becomes the same for both blocks. Since it's almost the end of my day, I haven't had time to investigate further, but I believe this is something we need to address before merging the PR.

Could you try to replicate the issue as well? Here is a quick demo:

https://github.com/woocommerce/woocommerce-blocks/assets/16707866/ee507bbc-579b-42fd-8520-ae2f91ebec13

**Comment by DAreRodz** [comment]
> Could you try to replicate the issue as well? Here is a quick demo:

Hey, @imanish003, I was able to reproduce the issue as well. My product catalog contains two Product Query blocks with different IDs (manually specified), and when I click `Next Page` on any of them, both become the second one.

Taking a deeper look right now. 🕵️

**Comment by DAreRodz** [comment]
I think I fixed it in https://github.com/woocommerce/woocommerce-blocks/pull/10200/commits/763ddf851a86879e84f1290eba40d9ea2beda2ba. It turned out the logic wasn't obtaining the right IDs from the `data-wc-navigation-id` attributes. Instead of doing

```js
const id = region.getAttribute( attrName );
```

it was doing

```js
const id = region.attributes[ attrName ];
```

which doesn't return the attribute's value but an [Attr](https://developer.mozilla.org/en-US/docs/Web/API/Attr) instance. This was later converted to a key with the text `[Object Attr]`, and then both vDOM trees were stored under that same key.

This is the behavior after the fix.


https://github.com/woocommerce/woocommerce-blocks/assets/6917969/d185d88f-24c9-4ddb-98ce-75251188df92

**Comment by imanish003** [comment]
Hey @DAreRodz, thank you so much for quickly making the changes. I tested the PR again, and it looks like the issue is fixed now. 🙌🏻

During the testing, I noticed that whenever I change the page, I am always taken to the top of the page. This might not be a good user experience, as the block can be anywhere on the page. Instead, maybe we should allow developers to provide a selector to the `data-wc-navigation-link` directive, and then scroll to that selector whenever the page changes. This way, we can scroll to the top of the provided selector instead of always going to the top of the page.

I tried to see how easy it would be to implement this idea, and I made an implementation in PR woocommerce/woocommerce-blocks#10349. Perhaps there is a better way to do it, but this is what I came up with at the moment.

Do you agree that instead of always going to the top of the page, we should allow developers to define a selector? I would love to hear your thoughts on this.

CC: @tjcafferkey @kmanijak @Aljullu 

Here is a quick demo of what I implemented in PR woocommerce/woocommerce-blocks#10349:


https://github.com/woocommerce/woocommerce-blocks/assets/16707866/02057d87-56b9-4e13-81c3-2bc748b75769

**Review by imanish003 — APPROVED** [review]
@DAreRodz thanks for quickly fixing it! 🙌🏻 I left one comment, but I don't think it's a blocker, so I am pre-approving the PR. 🙂

**Comment by kmanijak** [comment]
> Do you agree that instead of always going to the top of the page, we should allow developers to define a selector? I would love to hear your thoughts on this.

@imanish003, it's nice you touched on this topic! I like the idea to allow providing a custom selector to scroll to. But personally, I would change the default value: instead of scrolling to the top of the page, I think we should scroll to the top of the currently used Products/Product Collection block. My impression is that's the most common use case and that's what should be enabled by default.

And of course, if there's a need to scroll to another place, a custom selector could be provided as you suggested.

---

BTW, we could think of using [`scrollIntoView`](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView) which could simplify the scrolling logic.

**Comment by imanish003** [comment]
Hey @kmanijak,

> I would change the default value: instead of scrolling to the top of the page, I think we should scroll to the top of the currently used Products/Product Collection block.

I don't think we can modify the default behavior because as far as I know, the directive defined [here](https://github.com/woocommerce/woocommerce-blocks/blob/066376b7ad9683cfef0f108179de8cc68acc3746/assets/js/interactivity/directives.js#L151-L189) is intended to be more generic and shouldn't be aware of block-specific details, unless:
- This directive is a custom directive meant for use only with the Products beta & Product Collection blocks.

> BTW, we could think of using [scrollIntoView](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView) which could simplify the scrolling logic.

I had also considered using it, but I was concerned about older browser support. However, upon rechecking, it seems that `scrollIntoView` support is good, and we are already using it in our codebase. So, we can safely use it instead of `scrollTo`. Thanks for pointing this out. 🙌🏻

**Comment by luisherranz** [comment]
I'm not a UX expert, but I'd be careful not to move the scroll at all if the list is small and already fits the screen. In my experience, that's quite annoying.

Also, if the scroll needs to be moved, it'd be better to animate it (`smooth`), so the user can see in which direction is moving and doesn't get confused.

**Comment by Aljullu** [comment]
I agree with what has been mentioned so far:

1. If the entire product grid is visible on the page, I would avoid scrolling (I think `scrollIntoView({block:'nearest'})` already handles that).
2. If it's not, I would scroll to the top of the product grid, but I don't think we need to allow changing the selector.
3. And also +1 from my side to using `smooth`.

**Comment by Aljullu** [comment]
> Also, In [woocommerce-gutenberg-products-block.php](https://github.com/woocommerce/woocommerce-blocks/blob/8f1c4a314088d41d329d0441ecc0808214144b24/woocommerce-gutenberg-products-block.php), we have:
> 
> > Requires at least: 6.1

Huh, good catch! Actually this should be 6.2. We officially only support the latest WP version. :thinking: 

> But this PR uses `WP_HTML_Tag_Processor`, which was introduced in `6.2.0` as mentioned [here](https://developer.wordpress.org/reference/classes/wp_html_tag_processor/#changelog). I think we will need to do one of the following (unless I am missing something):
> 
> 1. Update `Requires at least` to `6.2` ,OR
> 2. Conditionally support the features introduced in this PR , OR
> 3. Figure out alternative to `WP_HTML_Tag_Processor` to update the $block_content

Once WP 6.3 is released in August, WC will update its supported versions to 6.2 or higher, so we should be good to go.

---

So, if I got the dates right, the _tl;dr_ is:
* for WC Blocks we need to do option 1 (Update `Requires at least` to `6.2`)
* for WC core we don't need to do anything

**Comment by tjcafferkey** [comment]
I think @Aljullu has summed it up nicely, especially if the following statement is true:

>If the entire product grid is visible on the page, I would avoid scrolling (I think scrollIntoView({block:'nearest'}) already handles that).

My assumption is that this behaviour will be most valuable to mobile users, but we like those have already said below me we shouldn't assume that the product grid is always going to extend beyond the height of the users screen.

>but I don't think we need to allow changing the selector

I think we should also avoid doing this.

**Comment by luisherranz** [comment]
> If the entire product grid is visible on the page, I would avoid scrolling (I think scrollIntoView({block:'nearest'}) already handles that).

It does, but it also scrolls to the bottom when the list is bigger than the screen.

I've tried both `nearest` and `start` with a Products list that's both smaller and bigger than the viewport, but none of them are perfect:

- `nearest` scrolls to the bottom when the list is bigger than the viewport. We want it to scroll to the top.
- `start` scrolls to the top when the list is smaller than the viewport. We want it to not scroll at all.

https://codepen.io/luisherranz/pen/ExOdjRK

For now, my opinion would be to go with `start`. Even though I find the scroll to the top annoying when the top is already in the viewport, I think it's worse to have to manually scroll to the top of the list each time you paginate, especially when the list is very long.

> I don't think we can modify the default behavior because as far as I know, the directive defined here is intended to be more generic and shouldn't be aware of block-specific details

Maybe we could try a new pattern for this: passing a callback.

```html
<a
  href="..."
  data-wc-navigation-link='{ "prefetch": true, "scroll": "ref:actions.woocommerce.scrollToProducts" }'
></a>
```

```js
store({
  actions: {
    woocommerce: {
      scrollToProducts: ({ ref }) => {
        ref.closest("[data-wc-navigation-id^='woo-products-'").scrollIntoView({
          behavior: "smooth",
          block: "start",
        });
      },
    },
  },
});
```

In the future, if what [Albert mentioned here](https://github.com/woocommerce/woocommerce-blocks/pull/10200#issuecomment-1651205386) is the desired UX, we could introduce an `Intersection Observer` and use `block: start` only when the top of the Products list is not on the viewport.

EDIT: I documented this pattern [here](https://github.com/WordPress/gutenberg/discussions/52973).

**Comment by Aljullu** [comment]
> It does, but it also scrolls to the bottom when the list is bigger than the screen.

Oh, I wasn't aware of that. Thanks for pointing it out!

> For now, my opinion would be to go with `start`. Even though I find the scroll to the top annoying when the top is already in the viewport, I think it's worse to have to manually scroll to the top of the list each time you paginate, especially when the list is very long.

It works for me. :+1:  We can always iterate on this in the future if needed.

**Comment by imanish003** [comment]
Hey @luisherranz,

> data-wc-navigation-link='{ "prefetch": true, "scroll": "ref:actions.woocommerce.scrollToProducts" }'

Thank you for sharing this and documenting this approach. This looks promising to me. I noticed that you have also commented on draft PR woocommerce/woocommerce-blocks#10349. I have a few questions regarding this approach, which I will ask on the PR 🙂

---

As far as I understand, it seems that what @Aljullu mentioned [here](https://github.com/woocommerce/woocommerce-blocks/pull/10200#issuecomment-1651205386) is the ideal behavior we would like to achieve. If so, should I continue working on draft PR woocommerce/woocommerce-blocks#10349 and see if this could be achieved using the approach shared by @luisherranz in the above comment? CC: @tjcafferkey

**Comment by imanish003** [comment]
I am going to bump the milestone of this PR as this doesn't have a blocker label.

**Comment by DAreRodz** [comment]
Hey, I'm just catching up with the conversation. I see a PR already opened (https://github.com/woocommerce/woocommerce-blocks/pull/10349) to explore the scroll behavior that was agreed upon (i.e., scroll smoothly to the top of the nearest Query block containing the link).

In any case, I guess we can merge this PR as it is, right? Or should we wait for the other PR to be ready? 🤔 There is also [this discussion](https://github.com/WordPress/gutenberg/discussions/52973) to figure out a better API for directives that could enable these use cases.

cc: @luisherranz @Aljullu @imanish003

**Comment by luisherranz** [comment]
I would disable the scroll for now (`scroll: false` in `data-wp-navigation-link`) which is a better UX than scrolling to the top, and I'd merge this. The other enhancement is not strictly necessary and can be done later.

**Comment by imanish003** [comment]
@luisherranz @DAreRodz Just to confirm, it looks like before merging this PR, we haven't disabled the scroll for now. That is intended, right?

I am confirming this because I will be implementing the same for the Product Collection block too.

**Comment by DAreRodz** [comment]
Oops! 🙈 Nope, it wasn't intended; nice catch @imanish003. I think I assumed it was disabled when I merged it... 😅

I can open a PR with a quick fix later today.

**Comment by imanish003** [comment]
Sounds great! Thanks, @DAreRodz 🙌🏻

**Comment by DAreRodz** [comment]
I almost forgot; here it is.

- https://github.com/woocommerce/woocommerce-blocks/pull/10522

File: src/BlockTypes/ProductQuery.php
Comment (luisherranz) [reply src/BlockTypes/ProductQuery.php:152]: @DAreRodz I've added `data-wc-key` in https://github.com/woocommerce/woocommerce-blocks/pull/10604 so we should use it here as well.

**Pull request merged**
